### PR TITLE
Combine draw and fill rectangle drawing operation

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?

This changes the `IDrawingContextImpl` to only have one method to draw a rectangle. The user-facing `DrawingContext` stays almost the same but has a new method to draw a stroked and filled rectangle at once.

## What is the current behavior?

Currently, two operations are needed to draw a stroked and filled rectangle.

## What is the updated/expected behavior with this PR?

Only one method is needed to draw a stroked and filled rectangle. This also reduces the count of nodes that are needed in the scene graph to draw a `Border` with `Background`.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues

#2538

### CODE REVIEW

File: src/Avalonia.Visuals/Media/DrawingContext.cs [hunk 1]
```diff
@@ -154,20 +178,6 @@ public void DrawText(IBrush foreground, Point origin, FormattedText text)
             }
         }
 
-        /// <summary>
-        /// Draws a filled rectangle.
-        /// </summary>
-        /// <param name="brush">The brush.</param>
-        /// <param name="rect">The rectangle bounds.</param>
-        /// <param name="cornerRadius">The corner radius.</param>
-        public void FillRectangle(IBrush brush, Rect rect, float cornerRadius = 0.0f)
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/DrawingContext.cs]: We shouldn't break backwards compatibility in our user-facing DrawingContext. Consider keeping previous versions.

File: src/Avalonia.Visuals/Media/DrawingContext.cs [hunk 2]
```diff
@@ -119,16 +119,40 @@ public void DrawGeometry(IBrush brush, IPen pen, Geometry geometry)
...
+        /// <param name="brush">The brush.</param>
+        /// <param name="pen">The pen.</param>
+        /// <param name="rect">The rectangle bounds.</param>
+        /// <param name="radiusX"></param>
+        /// <param name="radiusY"></param>
+        /// <remarks>
+        /// The brush and the pen can both be null. If the brush is null, then no fill is performed.
+        /// If the pen is null, then no stoke is performed. If both the pen and the brush are null, then the drawing is not visible.
+        /// </remarks>
+        public void DrawRoundedRectangle(IBrush brush, IPen pen, Rect rect, double radiusX, double radiusY)
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/DrawingContext.cs]: Why do we need a separate method instead of default argument value as we did before?

**Review by kekekeks — CHANGES_REQUESTED** [review]
PR also has several whitespace changes. Please, remove them and squash them out to avoid polluting `git blame` output.

File: src/Avalonia.Visuals/Media/DrawingContext.cs (hunk 2 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/DrawingContext.cs]: This just matches the Wpf API. We can use a single method with default parameters if you want.

**Comment by Gillibald** [comment]
How do we deal with these styling errors? (extra whitespaces etc.) Is someone willing to fix these separately or are keeping them forever?

**Comment by jmacato** [comment]
@Gillibald would be great if we could have a one-time enforcement of editorconfig rules repo-wide.. I keep having these silly whitespace/formatting mismatch issues and i dont wanna deal with that too much.

**Comment by jmacato** [comment]
And also we could add CodeFactor to automate editorconfig enforcement afterwards, like we do in Wasabi.

**Comment by Gillibald** [comment]
I will revert the whitespace changes and fix the code style in different PRs

**Comment by Gillibald** [comment]
I will revert the removal of some API and mark it obsolete. So people can transition to the new API within one release or so.

**Comment by kekekeks** [comment]
@jmacato one can also use some Git UI tool to actually preview what is being commited instead of pushing all changed files to the repo.

File: samples/ControlCatalog/Pages/PointersPage.cs [hunk 3]
```diff
@@ -95,7 +95,7 @@ protected override void OnPointerCaptureLost(PointerCaptureLostEventArgs e)
 
         public override void Render(DrawingContext context)
         {
-            context.FillRectangle(Brushes.Transparent, new Rect(default, Bounds.Size));
+            context.DrawRectangle(Brushes.Transparent, null, new Rect(default, Bounds.Size));
```
Comment (grokys) [inline samples/ControlCatalog/Pages/PointersPage.cs]: This change and the other changes in ControlCatalog can be reverted now, as we've kept the old API in place.

File: src/Avalonia.Controls/Presenters/TextPresenter.cs [hunk 4]
```diff
@@ -156,7 +156,7 @@ public override void Render(DrawingContext context)
 
                 foreach (var rect in rects)
                 {
-                    context.FillRectangle(SelectionBrush, rect);
+                    context.DrawRectangle(SelectionBrush, null, rect);
```
Comment (grokys) [inline src/Avalonia.Controls/Presenters/TextPresenter.cs]: Change can be reverted.

File: src/Avalonia.Controls/Primitives/OverlayPopupHost.cs [hunk 5]
```diff
@@ -143,7 +143,7 @@ public static IPopupHost CreatePopupHost(IVisual target, IAvaloniaDependencyReso
 
         public override void Render(DrawingContext context)
         {
-            context.FillRectangle(Brushes.White, new Rect(default, Bounds.Size));
+            context.DrawRectangle(Brushes.White, null, new Rect(default, Bounds.Size));
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/OverlayPopupHost.cs]: Change can be reverted.

File: src/Avalonia.Visuals/Media/DrawingContext.cs (hunk 2 above)
Comment (grokys) [reply src/Avalonia.Visuals/Media/DrawingContext.cs]: I'm undecided on this. I guess WPF compatibility is a useful thing to have, but how compatible are we on this API of this class in general?

**Review by grokys — CHANGES_REQUESTED** [review]
LGTM, but there's quite a lot of noise where calls to `DrawingContext.FillRectangle` have been changed to `DrawRectangle` (I marked a few of them but others remain), I think it would be better to revert these changes as they don't add anything.

**Comment by Gillibald** [comment]
@grokys @kekekeks
I think the question is do we need multiple methods to draw a rectangle? For me, it makes no sense to keep `FillRectangle`. We can just have one method `DrawRectangle(Brush, Pen, Rect, RadiusX, RadiusY)` old API should be marked obsolete and removed in a later release. To be able to mark the old API obsolete I changed the ControlCatalog to use the new API.

**Comment by kekekeks** [comment]
`DrawingContext` class is forwarding the call to `IDrawingContextImpl` *anyway*. So it doesn't hurt to have an extra method, IMO.

That is true for `DrawRoundedRectangle` for compatibility with copy-pasted WPF code and for keeping specialized overloads that only fill or stroke.

@grokys @Gillibald

**Comment by grokys** [comment]
Yeah I agree. If we were designing this API from scratch we'd only have one method, but one of our goals is keeping compatibility with WPF where it makes sense moving forwards. Given that they're all forwarding the a single call in the platform implementation, I think it's a good idea to keep the WPF-style overloads.

File: src/Avalonia.Visuals/Media/DrawingContext.cs (hunk 2 above)
Comment (Gillibald) [reply src/Avalonia.Visuals/Media/DrawingContext.cs]: DrawRectangle now has the same signature just a different name.

**Review by grokys — APPROVED** [review]
LGTM now, thanks @Gillibald !

**Review by kekekeks — APPROVED** [review]
There are quite a lot of lines where `, null` is added while they are still compatible with the old API. Not sure if it's OK or not, so feel free to keep the PR as is.

Other than that, LGTM

**Comment by Gillibald** [comment]
The additions where need because at these places the `IDrawingContextImpl` interface was used if I am not mistaken.

File: src/Skia/Avalonia.Skia/DrawingContextImpl.cs [hunk 6]
```diff
@@ -184,37 +184,40 @@ public void DrawGeometry(IBrush brush, IPen pen, IGeometryImpl geometry)
...
 
         /// <inheritdoc />
-        public void DrawRectangle(IPen pen, Rect rect, float cornerRadius = 0)
+        public void DrawRectangle(IBrush brush, IPen pen, Rect rect, double radiusX, double radiusY)
         {
-            using (var paint = CreatePaint(pen, rect.Size))
-            {
-                var rc = rect.ToSKRect();
+            var rc = rect.ToSKRect();
+            var isRounded = Math.Abs(radiusX) > double.Epsilon || Math.Abs(radiusX) > double.Epsilon;
```
Comment (Kermalis) [inline src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: I believe here you meant to check radiusX and radiusY, but you checked radiusX twice.

Comment (Kermalis) [reply src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: Same problem in `Windows\Avalonia.Direct2D1\Media\DrawingContextImpl.cs`

Comment (Gillibald) [reply src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: Y. Will fix asap.

Comment (Gillibald) [reply src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: #3324

Comment (grokys) [reply src/Skia/Avalonia.Skia/DrawingContextImpl.cs]: Sharp eyes @Kermalis ! 👀

**Pull request merged**